File: countunivaluesubtrees.js
//Objective is to find the number of unique univalue subtrees in a binary tree.

class Node {
    constructor(left, right, val) {
      this.left = left
      this.right = right
      this.val = val
    }
}
  
class Tree {
    constructor(root) {
      this.root = null
    }
  
    createRoot(val) {
      this.root = new Node(null, null, val)
    }
  
    addLeftNode(node, val) {
      node.left = new Node(null, null, val)
    }
  
    addRightNode(node, val) {
      node.right = new Node(null, null, val)
    }
}

let tree = new Tree()
tree.createRoot(5)
tree.addLeftNode(tree.root, 1)
tree.addRightNode(tree.root, 5)
tree.addLeftNode(tree.root.left, 5)
tree.addRightNode(tree.root.right, 5)
tree.addRightNode(tree.root.left, 5)


//O(n) solution that does a dfs of the tree and checks every node and its children for unnique subtrees.

let count = 0

function dfs(node) {
    if (!node) {
        return
    }

    if (isUnival(node, node.val)) {
        count++
    }

    dfs(node.left)
    dfs(node.right)
}
dfs(tree.root)

function isUnival(node, value) {
    if (!node) {
        return true
    }

    if (!node.left && !node.right && node.val === value) {
        return true
    }

    return node.val === value && isUnival(node.left, node.val) && isUnival(node.right, node.val)
}

return count

File: countunivaluesubtrees.java
//Java Solution

class Solution {
    public int countUnivalSubtrees(TreeNode root) {
        int count = 0;
        count = dfs(root, count);
        return count;
    }
    
    private int dfs(TreeNode node, Integer count) {
        if (node == null) {
            return count;
        }
        
        if (isUnival(node, node.val)) {
            count++;
        }
        
        count = dfs(node.left, count);
        count = dfs(node.right, count);
        
        return count;
    }
    
    private boolean isUnival(TreeNode node, Integer value) {
        if (node == null) {
            return true;
        }
        
        if (node.left == null && node.right == null && node.val == value) {
            return true;
        }
        
        return node.val == value && isUnival(node.left, value) && isUnival(node.right, value);
    }
}